### ISSUE TITLE
Support for dynamic repository configuration

### ISSUE THREAD

**Issue by devversion** [renovatebot/renovate#21110]
### What would you like Renovate to be able to do?

I would like to configure renovate in my repository using a dynamically-created repository config. 

- e.g `renovate.json` could be replaced with `renovate-config.mjs` that would be evaluated using NodeJS.  
- Discussion: https://github.com/renovatebot/renovate/discussions/20867

**Use case**

In my repository, we have branches for every minor version that we publish. e.g. `14.2.x`, `14.3.x`, `15.0.x`. In a simplified manner, we would like Renovate to only operate on the two most highest SemVer branches. This is currently not possible to achieve via a RegExp.

There are ways to achieve it by manually disabling Renovate for these older branches, but we would **like to avoid** manually having to maintaining this list/landing changes in older branches.

We have a script that will tell us the most recent branches that are "still in active development" and would like to use it to configure Renovate. Also with a dynamic configuration, we can have different configurations for different semantic branches. e.g.

```js
- main   // should use active @next version of Angular
- 15.2.x // should active @latest version of Angular
- 15.1.x // no longer in active development. Would not like Renovate PRs
- 15.0.x // no longer in active development. Would not like Renovate PRs
```

### If you have any ideas on how this should be implemented, please tell us here.

Mostly thinking of supporting a `.js`, `.mjs`, `.cjs` config file in a repository. Renovate would then run it via Node. It would be opt-in only for self-hosted users. The configuration file *could* not depend on any `node_modules`?

### Is this a feature you are interested in implementing yourself?

Maybe

**Comment by RahulGautamSingh** [comment]
Is it possible to use preset for this case, where we can automatically update the `baseBranches` option in the preset using a script? (will help exclude outdated branches) 
I am not sure how we will generate config dynamically using a script. Very interested in the solutions for this one👀

**Comment by rarkins** [comment]
So in short,
- It should have a global config option `allowDynamicRepoConfig` defaulting to `false`
- It should look for `renovate.js/mjs/cjs` and adapt as necessary based on file extension

Not too much else to describe I think?

**Comment by rarkins** [comment]
That might be best handled via something new like a `postCloneCommands` or `postCloneTasks` array, with controls similar to postUpgradeTasks

**Comment by devversion** [comment]
good idea. That seems like a nice follow-up if we have the base functionality.

**Comment by HonkingGoose** [comment]
I think we're `status:ready` based on this scope:

> So in short,
> 
>     * It should have a global config option `allowDynamicRepoConfig` defaulting to `false`
> 
>     * It should look for `renovate.js/mjs/cjs` and adapt as necessary based on file extension

---

And I created #23963 for this idea:

> Yes. One follow-up question/point would be: could top-level `node_modules` be made available?

**Comment by matzeeable** [comment]
I just stumbled over this issue as I tried to create a workaround solution for https://github.com/renovatebot/renovate/discussions/29424. In general, I want to create a `renovate.json` from a dynamic mechanism like `renovate.mjs`. I would like to understand the coding of Renovate, so I could potentially create a PR, so would the following changes resolve the issue?

1. Add `renovate.mjs` to

https://github.com/renovatebot/renovate/blob/4039ace0d1f30b7c0aff64dc74da5cee582308aa/lib/config/app-strings.ts#L1-L12

2. When it comes to reading the config, it runs

https://github.com/renovatebot/renovate/blob/4039ace0d1f30b7c0aff64dc74da5cee582308aa/lib/config-validator.ts#L90-L97

3. Add a `switch` case for `mjs` and `import()` the module

https://github.com/renovatebot/renovate/blob/4039ace0d1f30b7c0aff64dc74da5cee582308aa/lib/workers/global/config/parse/file.ts#L12-L27

**Comment by rarkins** [comment]
It's more complicated than you describe because of the need to control whether it's allowed or not. For many or most Renovate admins, they won't want to allow arbitrary script execution by repo users.
